{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find ortholog\n",
    "This script is witten to bla bla bla.. something orthologs\n",
    "\n",
    "Major clock genes in N. crassa with Ophio_kim orthologs\n",
    "* frq = Ophio5|6064\n",
    "* wc-1 = Ophio5|4975\n",
    "* wc-2 = Ophio5|889\n",
    "* vvd = Ophio5|6595\n",
    "* nik-2 = Ophio5|6786\n",
    "* dcc-1 = Ophio5|7010\n",
    "* luxQ = Ophio5|7293\n",
    "* phy-1 = Ophio5|4324\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Housekeeping\n",
    "\n",
    "import pandas as pd\n",
    "import sqlite3\n",
    "import numpy as np\n",
    "\n",
    "path = '/Users/roos_brouns/Dropbox/Ant-fungus/02_scripts/Git_Das_folder2/Das_et_al_2022a'\n",
    "species = 'ophio_cflo'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in file with ophio cflo and kim orthologs \n",
    "orthos = pd.read_csv(f'{path}/data/will_et_al_2020/FullBlast_EC05_RNAseq_orignal_copy_26Aug19.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the orthologs of clock genes in ophio_cflo\n",
    "\n",
    "orthos_genes = 'Ophio5|6064 Ophio5|4975 Ophio5|889 Ophio5|6595 Ophio5|6786 Ophio5|7010 Ophio5|7293 Ophio5|4324'\n",
    "\n",
    "list_w_orthos = orthos_genes.split()\n",
    "\n",
    "df = orthos.loc[orthos['sc16a_gene'] == list_w_orthos[0]]\n",
    "\n",
    "for no in range(1,len(list_w_orthos)):\n",
    "    df_line = orthos.loc[orthos['sc16a_gene'] == list_w_orthos[no]]\n",
    "    # print(df_line)\n",
    "    df = df.append(df_line, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Ophcf2|06690',\n",
       " 'Ophcf2|01775',\n",
       " 'Ophcf2|02346',\n",
       " 'Ophcf2|01187',\n",
       " 'Ophcf2|05709',\n",
       " 'Ophcf2|02284',\n",
       " 'Ophcf2|03331',\n",
       " 'Ophcf2|01137']"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kim_cflo = df[['arb2_gene','sc16a_gene']].drop_duplicates()\n",
    "kim_cflo.reset_index(inplace=True, drop=True)\n",
    "list(kim_cflo['arb2_gene'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID_ncbi</th>\n",
       "      <th>gene_ID_robin</th>\n",
       "      <th>rhythmic_24h</th>\n",
       "      <th>GammaP_24h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GQ602_006690</td>\n",
       "      <td>Ophcf2|06690</td>\n",
       "      <td>no</td>\n",
       "      <td>0.052462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GQ602_006690</td>\n",
       "      <td>Ophcf2|06690</td>\n",
       "      <td>no</td>\n",
       "      <td>0.052462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GQ602_006690</td>\n",
       "      <td>Ophcf2|06690</td>\n",
       "      <td>no</td>\n",
       "      <td>0.052462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GQ602_001775</td>\n",
       "      <td>Ophcf2|01775</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.021846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GQ602_001775</td>\n",
       "      <td>Ophcf2|01775</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.021846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GQ602_001775</td>\n",
       "      <td>Ophcf2|01775</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.021846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GQ602_002346</td>\n",
       "      <td>Ophcf2|02346</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.013689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GQ602_002346</td>\n",
       "      <td>Ophcf2|02346</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.013689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GQ602_002346</td>\n",
       "      <td>Ophcf2|02346</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.013689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GQ602_001187</td>\n",
       "      <td>Ophcf2|01187</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.002997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GQ602_001187</td>\n",
       "      <td>Ophcf2|01187</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.002997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GQ602_001187</td>\n",
       "      <td>Ophcf2|01187</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.002997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GQ602_005709</td>\n",
       "      <td>Ophcf2|05709</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.013689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GQ602_005709</td>\n",
       "      <td>Ophcf2|05709</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.013689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GQ602_005709</td>\n",
       "      <td>Ophcf2|05709</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.013689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GQ602_002284</td>\n",
       "      <td>Ophcf2|02284</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.001787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GQ602_002284</td>\n",
       "      <td>Ophcf2|02284</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.001787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GQ602_002284</td>\n",
       "      <td>Ophcf2|02284</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.001787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GQ602_003331</td>\n",
       "      <td>Ophcf2|03331</td>\n",
       "      <td>no</td>\n",
       "      <td>0.723429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GQ602_003331</td>\n",
       "      <td>Ophcf2|03331</td>\n",
       "      <td>no</td>\n",
       "      <td>0.723429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GQ602_003331</td>\n",
       "      <td>Ophcf2|03331</td>\n",
       "      <td>no</td>\n",
       "      <td>0.723429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>GQ602_001137</td>\n",
       "      <td>Ophcf2|01137</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.005102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GQ602_001137</td>\n",
       "      <td>Ophcf2|01137</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.005102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GQ602_001137</td>\n",
       "      <td>Ophcf2|01137</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.005102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_ID_ncbi gene_ID_robin rhythmic_24h  GammaP_24h\n",
       "0   GQ602_006690  Ophcf2|06690           no    0.052462\n",
       "1   GQ602_006690  Ophcf2|06690           no    0.052462\n",
       "2   GQ602_006690  Ophcf2|06690           no    0.052462\n",
       "3   GQ602_001775  Ophcf2|01775          yes    0.021846\n",
       "4   GQ602_001775  Ophcf2|01775          yes    0.021846\n",
       "5   GQ602_001775  Ophcf2|01775          yes    0.021846\n",
       "6   GQ602_002346  Ophcf2|02346          yes    0.013689\n",
       "7   GQ602_002346  Ophcf2|02346          yes    0.013689\n",
       "8   GQ602_002346  Ophcf2|02346          yes    0.013689\n",
       "9   GQ602_001187  Ophcf2|01187          yes    0.002997\n",
       "10  GQ602_001187  Ophcf2|01187          yes    0.002997\n",
       "11  GQ602_001187  Ophcf2|01187          yes    0.002997\n",
       "12  GQ602_005709  Ophcf2|05709          yes    0.013689\n",
       "13  GQ602_005709  Ophcf2|05709          yes    0.013689\n",
       "14  GQ602_005709  Ophcf2|05709          yes    0.013689\n",
       "15  GQ602_002284  Ophcf2|02284          yes    0.001787\n",
       "16  GQ602_002284  Ophcf2|02284          yes    0.001787\n",
       "17  GQ602_002284  Ophcf2|02284          yes    0.001787\n",
       "18  GQ602_003331  Ophcf2|03331           no    0.723429\n",
       "19  GQ602_003331  Ophcf2|03331           no    0.723429\n",
       "20  GQ602_003331  Ophcf2|03331           no    0.723429\n",
       "21  GQ602_001137  Ophcf2|01137          yes    0.005102\n",
       "22  GQ602_001137  Ophcf2|01137          yes    0.005102\n",
       "23  GQ602_001137  Ophcf2|01137          yes    0.005102"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Are these orthologs also rhythmic?\n",
    "\n",
    "species='ophio_cflo'\n",
    "\n",
    "# put orthologs of clock genes in a list\n",
    "cflo_orhts = df['arb2_gene']\n",
    "cflo_orhts = list(cflo_orhts)\n",
    "cflo_orhts[0]\n",
    "\n",
    "# load in data with info if gene is rhytmic or not\n",
    "#\n",
    "# load in data with info if gene is rhytmic or not\n",
    "#\n",
    "# Load in the whole csv\n",
    "data = pd.read_csv(f'{path}/data/{species}_TC6_data.csv')\n",
    "# select the expression values\n",
    "\n",
    "# data.columndata.column's\n",
    "\n",
    "rhytmic_genes = data.filter(['gene_ID_ncbi','gene_ID_robin', 'rhythmic_24h', 'GammaP_24h'])\n",
    "\n",
    "## make df with the genes that are rhytmic \n",
    "rhy_df = rhytmic_genes.loc[rhytmic_genes['gene_ID_robin'] == cflo_orhts[0]]\n",
    "\n",
    "for i in range(1,len(cflo_orhts)):\n",
    "    rhy_df2 = rhytmic_genes.loc[rhytmic_genes['gene_ID_robin'] == cflo_orhts[i]]\n",
    "    rhy_df = rhy_df.append(rhy_df2, ignore_index = True)\n",
    "\n",
    "rhy_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Are these orthologs also rhythmic in Beau?\n",
    "species='beau'\n",
    "\n",
    "# put orthologs of clock genes in a list\n",
    "orhts = ['BBA_01528', 'BBA_10271', 'BBA_01403', 'BBA_02876', 'BBA_08737', 'BBA_00328', 'BBA_07925', 'BBA_02816']\n",
    "\n",
    "# load in data with info if gene is rhytmic or not\n",
    "#\n",
    "# Load in the whole csv\n",
    "data = pd.read_csv(f'{path}/data/{species}_TC6_data.csv')\n",
    "# select the expression values\n",
    "\n",
    "# data.columndata.column's\n",
    "\n",
    "b_rhytmic_genes = data.filter(['gene_ID_ncbi', 'rhythmic_24h', 'GammaP_24h'])\n",
    "\n",
    "\n",
    "## make df with the genes that are rhytmic \n",
    "b_rhy_df = b_rhytmic_genes.loc[b_rhytmic_genes['gene_ID_ncbi'] == orhts[0]]\n",
    "\n",
    "for i in range(1,len(orhts)):\n",
    "    b_rhy_df2 = b_rhytmic_genes.loc[b_rhytmic_genes['gene_ID_ncbi'] == orhts[i]]\n",
    "    b_rhy_df = b_rhy_df.append(b_rhy_df2, ignore_index = True)\n",
    "\n",
    "b_rhydf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ophio_cflo_ncbi</th>\n",
       "      <th>ophio_cflo_robin_ID</th>\n",
       "      <th>rhythmic_cflo</th>\n",
       "      <th>GammaP_cflo</th>\n",
       "      <th>ophio_kim_ortho</th>\n",
       "      <th>beau_ortho</th>\n",
       "      <th>rhythmic_beau</th>\n",
       "      <th>GammaP_beau</th>\n",
       "      <th>N_crassa_ortho</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GQ602_006690</td>\n",
       "      <td>Ophcf2|06690</td>\n",
       "      <td>no</td>\n",
       "      <td>0.052</td>\n",
       "      <td>Ophio5|6064</td>\n",
       "      <td>BBA_01528</td>\n",
       "      <td>no</td>\n",
       "      <td>0.082</td>\n",
       "      <td>frq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GQ602_001775</td>\n",
       "      <td>Ophcf2|01775</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.022</td>\n",
       "      <td>Ophio5|4975</td>\n",
       "      <td>BBA_10271</td>\n",
       "      <td>no</td>\n",
       "      <td>0.119</td>\n",
       "      <td>wc-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GQ602_002346</td>\n",
       "      <td>Ophcf2|02346</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.014</td>\n",
       "      <td>Ophio5|889</td>\n",
       "      <td>BBA_01403</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.012</td>\n",
       "      <td>wc-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GQ602_001187</td>\n",
       "      <td>Ophcf2|01187</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.003</td>\n",
       "      <td>Ophio5|6595</td>\n",
       "      <td>BBA_02876</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.037</td>\n",
       "      <td>vvd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GQ602_005709</td>\n",
       "      <td>Ophcf2|05709</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.014</td>\n",
       "      <td>Ophio5|6786</td>\n",
       "      <td>BBA_08737</td>\n",
       "      <td>no</td>\n",
       "      <td>0.055</td>\n",
       "      <td>nik-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GQ602_002284</td>\n",
       "      <td>Ophcf2|02284</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.002</td>\n",
       "      <td>Ophio5|7010</td>\n",
       "      <td>BBA_00328</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.019</td>\n",
       "      <td>dcc-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GQ602_003331</td>\n",
       "      <td>Ophcf2|03331</td>\n",
       "      <td>no</td>\n",
       "      <td>0.723</td>\n",
       "      <td>Ophio5|7293</td>\n",
       "      <td>BBA_07925</td>\n",
       "      <td>no</td>\n",
       "      <td>0.310</td>\n",
       "      <td>luxQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GQ602_001137</td>\n",
       "      <td>Ophcf2|01137</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.005</td>\n",
       "      <td>Ophio5|4324</td>\n",
       "      <td>BBA_02816</td>\n",
       "      <td>no</td>\n",
       "      <td>0.506</td>\n",
       "      <td>phy-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ophio_cflo_ncbi ophio_cflo_robin_ID rhythmic_cflo  GammaP_cflo  \\\n",
       "0    GQ602_006690        Ophcf2|06690            no        0.052   \n",
       "1    GQ602_001775        Ophcf2|01775           yes        0.022   \n",
       "2    GQ602_002346        Ophcf2|02346           yes        0.014   \n",
       "3    GQ602_001187        Ophcf2|01187           yes        0.003   \n",
       "4    GQ602_005709        Ophcf2|05709           yes        0.014   \n",
       "5    GQ602_002284        Ophcf2|02284           yes        0.002   \n",
       "6    GQ602_003331        Ophcf2|03331            no        0.723   \n",
       "7    GQ602_001137        Ophcf2|01137           yes        0.005   \n",
       "\n",
       "  ophio_kim_ortho beau_ortho rhythmic_beau  GammaP_beau N_crassa_ortho  \n",
       "0     Ophio5|6064  BBA_01528            no        0.082            frq  \n",
       "1     Ophio5|4975  BBA_10271            no        0.119           wc-1  \n",
       "2      Ophio5|889  BBA_01403           yes        0.012           wc-2  \n",
       "3     Ophio5|6595  BBA_02876           yes        0.037            vvd  \n",
       "4     Ophio5|6786  BBA_08737            no        0.055          nik-2  \n",
       "5     Ophio5|7010  BBA_00328           yes        0.019          dcc-1  \n",
       "6     Ophio5|7293  BBA_07925            no        0.310           luxQ  \n",
       "7     Ophio5|4324  BBA_02816            no        0.506          phy-1  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Rename columns and make df pretty\n",
    "rhy_df.drop_duplicates(inplace=True)\n",
    "rhy_df.reset_index(inplace=True, drop=True)\n",
    "rhy_df.columns = ['ophio_cflo_ncbi','ophio_cflo_robin_ID', 'rhythmic_cflo', 'GammaP_cflo']\n",
    "\n",
    "\n",
    "# Add ophio_kim orthologs\n",
    "n_df = pd.concat([rhy_df, kim_cflo], join='outer', axis=1)\n",
    "# Ad beau orthologs\n",
    "n_df = pd.concat([n_df, b_rhy_df], join='outer', axis=1)\n",
    "# Ad N crassa orthologs\n",
    "n_df['N_crassa_ortho']= ['frq','wc-1','wc-2','vvd','nik-2','dcc-1','luxQ','phy-1']\n",
    "\n",
    "## Rename columns and make df pretty\n",
    "n_df.drop('arb2_gene', axis=1, inplace=True)\n",
    "n_df.columns = ['ophio_cflo_ncbi','ophio_cflo_robin_ID', 'rhythmic_cflo', 'GammaP_cflo', 'ophio_kim_ortho', 'beau_ortho','rhythmic_beau', 'GammaP_beau', 'N_crassa_ortho']\n",
    "# round decimals\n",
    "n_df.round(3)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ad N crassa orthologs\n",
    "n_df['N_crassa_ortho']= ['frq','wc-1','wc-2','vvd','nik-2','dcc-1','luxQ','phy-1']\n",
    "n_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_df.columns = ['ophio_cflo_ncbi','ophio_cflo_robin_ID', 'rhythmic_cflo', 'GammaP_cflo', 'ophio_kim_ortho']\n",
    "n_df['Beau_ortho']=['BBA_01528', 'BBA_10271', 'BBA_01403', 'BBA_02876', 'BBA_08737', 'BBA_00328', 'BBA_07925', 'BBA_02816']\n",
    "n_df['N_crassa_ortho']= ['frq','wc-1','wc-2','vvd','nik-2','dcc-1','luxQ','phy-1']\n",
    "n_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Are my genes rhythmic?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID_ncbi</th>\n",
       "      <th>rhythmic_24h</th>\n",
       "      <th>GammaP_24h</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBA_01528</td>\n",
       "      <td>no</td>\n",
       "      <td>0.081892</td>\n",
       "      <td>frq1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBA_08957</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>frq2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBA_02496</td>\n",
       "      <td>no</td>\n",
       "      <td>0.118621</td>\n",
       "      <td>frh</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_ID_ncbi rhythmic_24h  GammaP_24h gene_name\n",
       "0    BBA_01528           no    0.081892      frq1\n",
       "1    BBA_08957          NaN         NaN      frq2\n",
       "2    BBA_02496           no    0.118621       frh"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Are these orthologs also rhythmic in Beau?\n",
    "species='beau'\n",
    "gene_ID = 'gene_ID_ncbi'\n",
    "\n",
    "# input list of genes you want to test\n",
    "orhts = ['BBA_01528', 'BBA_08957', 'BBA_02496']\n",
    "\n",
    "# load in data with info if gene is rhytmic or not\n",
    "#\n",
    "# Load in the whole csv\n",
    "data = pd.read_csv(f'{path}/data/{species}_TC6_data.csv')\n",
    "# select the expression values\n",
    "\n",
    "# data.columndata.column's\n",
    "\n",
    "t_rhytmic_genes = data.filter([gene_ID, 'rhythmic_24h', 'GammaP_24h'])\n",
    "\n",
    "\n",
    "## make df with the genes that are rhytmic \n",
    "t_rhy_df = t_rhytmic_genes.loc[t_rhytmic_genes[gene_ID] == orhts[0]]\n",
    "\n",
    "for i in range(1,len(orhts)):\n",
    "    t_rhy_df2 = t_rhytmic_genes.loc[t_rhytmic_genes[gene_ID] == orhts[i]]\n",
    "    t_rhy_df = t_rhy_df.append(t_rhy_df2, ignore_index = True)\n",
    "\n",
    "    \n",
    "# ad gene names\n",
    "t_rhy_df['gene_name']= ['frq1','frq2','frh']\n",
    "\n",
    "\n",
    "t_rhy_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
