---
title: "compile_gene_names_from_gff_files"
author: "Biplabendu Das"
date: "8/17/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

# SPECIES: Ophiocordyceps kimflemingae

Read the two gff files - one from Robin and the other from NCBI

```{r}
ophio_kim_gff_robin <- read.csv("/Users/biplabendudas/Documents/GitHub/Das_et_al_2022a/data/input/ophio_kim/gffs/files_from_Robin/Ophio5.gene_features.gff3.csv",
                                header = T, stringsAsFactors = F)

ophio_kim_gff_ncbi <- read.csv("/Users/biplabendudas/Documents/GitHub/Das_et_al_2022a/data/input/ophio_kim/gffs/files_from_NCBI/GCA_001272575.2_ASM127257v2_genomic.gff.csv",
                               header = T, stringsAsFactors = F)

```

Step one: 
- Filter and keep only the rows with feature=="gene"
- Select only the "start", "end", and "attributes" column

```{r}

# load libraries
library(tidyverse)

ophio_kim_gff_robin <- ophio_kim_gff_robin %>% 
  filter(feature == "gene") %>% 
  select(start, end, attributes_robin = attributes)

ophio_kim_gff_ncbi <- ophio_kim_gff_ncbi %>% 
  filter(feature == "gene") %>% 
  select(start, end, attributes_ncbi = attributes)

```

Check to see if your code worked!

```{r}

head(ophio_kim_gff_robin)
head(ophio_kim_gff_ncbi)

```

Step two:
- Clean the attributes column to retain only the gene names

```{r}

## Robin's annotation

  # initial split
  foo <- ophio_kim_gff_robin %>% 
    separate(attributes_robin, c("a","b"), "=", extra = "merge")
  
  # final split to obtain only the gene names
  ophio_kim_gff_robin$attributes_robin <- unlist(strsplit(foo$b,"|gene",4))
  
## NCBI's annotation
  
  # all-in-one
  ophio_kim_gff_ncbi <- 
    ophio_kim_gff_ncbi %>% 
      separate(attributes_ncbi, c("a","b","c"), ";", extra = "merge") %>% 
      select(-c(a,c)) %>% 
      separate(b, c("b","c"), "=", extra = "merge") %>% 
      select(start, end, attributes_ncbi=c)
  
```

Step three:
- Merge the two files based on the "start" and "end" columns

```{r}

ophio_kim_gff_robin_ncbi <- 
  ophio_kim_gff_robin %>% 
    full_join(ophio_kim_gff_ncbi)

# SAVE THE FILE
write.csv(ophio_kim_gff_robin_ncbi, 
          file = "/Users/biplabendudas/Documents/GitHub/Das_et_al_2022a/results/normalized_gene_exp/ophio_kim/ophio_kim_gene_names_robin_ncbi.csv",
          row.names = F)

```


# SPECIES: Beauveria bassiana (but also inter-change-able)

Name your species and database name of Robin and NCBI, in this case beau, along with the path to the file
```{r}
species_name <- 'beau'
DB_name_Robin <- 'Beaba1'
DB_name_NCBI <- 'ARSEF2860'
path_R <- paste0('./Das_et_al_2022a/data/input/',{species_name},'/', {DB_name_Robin},'.genes.gff3')
path_NCBI <- paste0('./Das_et_al_2022a/data/input/',{species_name},'/',{species_name},'_',{DB_name_NCBI},'_genome.gff')
```


Read the two gff files 
* one from Robin (can be downloaded at: .. ) 

```{r}
gff_robin <- read.csv(path_R, sep = '\t', header = F, stringsAsFactors = F)
```

* the other from NCBI (can be downloaded)
This file has to be edited to work. The lines that start with a '#' are removed. This can be done with the following command line code (bash) or manually

```bash
file='give_file_name'
awk -F '\t' -v OFS=, '!/^#/ {$1=$1;print}' ${file}.gff > ${file}.csv;
```
Then read the file in

```{r}
gff_ncbi <- read.csv(path_NCBI, sep = '\t', header = F, stringsAsFactors = F)

```
And name the columns

```{r}
col_names <- c('seqname','source','feature','start','end','score','strand','frame','attributes')
colnames(gff_robin) <- col_names
colnames(gff_ncbi) <- col_names

```

Step one: 
- Filter and keep only the rows with feature=="gene"
- Select only the "start", "end", and "attributes" column

```{r}

# load libraries
library(tidyverse)

gff_robin <- gff_robin %>% 
  filter(feature == "gene") %>% 
  select(start, end, attributes_robin = attributes)

gff_ncbi <- gff_ncbi %>% 
  filter(feature == "gene") %>% 
  select(start, end, attributes_ncbi = attributes)

```

Check to see if your code worked!

```{r}

head(gff_robin)
head(gff_ncbi)

```

Step two:
- Clean the attributes column to retain only the gene names

```{r}

## Robin's annotation

  # initial split
  foo <- gff_robin %>% 
    separate(attributes_robin, c("a","b"), "=", extra = "merge")
  
  # final split to obtain only the gene names
  gff_robin$attributes_robin <- unlist(strsplit(foo$b,"|gene",4))
  
## NCBI's annotation
  
  # all-in-one
  gff_ncbi <-
    gff_ncbi %>% 
      separate(attributes_ncbi, c("a","b","c"), ";", extra = "merge") %>% 
      select(-c(a,c)) %>% 
      separate(b, c("b","c"), "=", extra = "merge") %>% 
      select(start, end, attributes_ncbi=c)
  
```

Step three:
- Merge the two files based on the "start" and "end" columns

```{r}

gff_robin_ncbi <- 
  gff_robin %>% 
    full_join(gff_ncbi)

# SAVE THE FILE
file_name <- paste0('./Das_et_al_2022a/data/input/',{species_name},'/',{species_name},'_gene_names_robin_ncbi.csv')
write.csv(gff_robin_ncbi, 
          file = file_name,
          row.names = F)

```


# SPECIES: Ophiocordyceps camponoti-floridani

Read the two gff files - one from Robin and the other from NCBI

```{r}
ophio_cflo_gff_robin <- read.csv("/Users/biplabendudas/Documents/GitHub/Das_et_al_2022a/data/input/ophio_cflo/gffs/files_from_robin/Ophcf2.gene_features.gff3.csv",
                                header = T, stringsAsFactors = F)

ophio_cflo_gff_ncbi <- read.csv("/Users/biplabendudas/Documents/GitHub/Das_et_al_2022a/data/input/ophio_cflo/gffs/files_from_ncbi/GCA_012980515.1_Ophcf2_genomic.gff.csv",
                               header = T, stringsAsFactors = F)

```

Step one: 
- Filter and keep only the rows with feature=="gene"
- Select only the "start", "end", and "attributes" column

```{r}

# load libraries
library(tidyverse)

ophio_cflo_gff_robin <- ophio_cflo_gff_robin %>% 
  filter(feature == "gene") %>% 
  select(start, end, attributes_robin = attributes)

ophio_cflo_gff_ncbi <- ophio_cflo_gff_ncbi %>% 
  filter(feature == "gene") %>% 
  select(start, end, attributes_ncbi = attributes)

```

Check to see if your code worked!

```{r}

head(ophio_cflo_gff_robin)
head(ophio_cflo_gff_ncbi)

```

Step two:
- Clean the attributes column to retain only the gene names

```{r}

## Robin's annotation

  # initial split
  foo <- ophio_cflo_gff_robin %>% 
    separate(attributes_robin, c("a","b"), "=", extra = "merge")
  
  # final split to obtain only the gene names
  ophio_cflo_gff_robin$attributes_robin <- unlist(strsplit(foo$b,"|gene",4))
  
## NCBI's annotation
  
  # all-in-one
  ophio_cflo_gff_ncbi <-
    ophio_cflo_gff_ncbi %>% 
      separate(attributes_ncbi, c("a","b","c"), ";", extra = "merge") %>% 
      select(-c(a,c)) %>% 
      separate(b, c("b","c"), "=", extra = "merge") %>% 
      select(start, end, attributes_ncbi=c)
  
```

Step three:
- Merge the two files based on the "start" and "end" columns

```{r}

ophio_cflo_gff_robin_ncbi <- 
  ophio_cflo_gff_robin %>% 
    full_join(ophio_cflo_gff_ncbi)

# SAVE THE FILE
write.csv(ophio_cflo_gff_robin_ncbi, 
          file = "/Users/biplabendudas/Documents/GitHub/Das_et_al_2022a/data/input/ophio_cflo/gffs/ophio_cflo_gene_names_robin_ncbi.csv",
          row.names = F)

```
